File: greenhouse_aiot/backend/models/prediction.py
"""Prediction model — maps to the `predictions` table defined in D1 schema.sql."""

from datetime import datetime

from models import db


class Prediction(db.Model):
    """AI model inference result stored for traceability."""

    __tablename__ = "predictions"

    prediction_id     = db.Column(db.Integer,    primary_key=True)
    device_id         = db.Column(db.Integer,    db.ForeignKey("devices.device_id"),          nullable=False)
    reading_id        = db.Column(db.BigInteger, db.ForeignKey("sensor_readings.reading_id"), nullable=False)
    model_name        = db.Column(db.String(100), nullable=False)
    model_version     = db.Column(db.String(20),  nullable=False)
    predicted_class   = db.Column(db.String(20),  nullable=False)
    confidence        = db.Column(db.Numeric(5, 4), nullable=False)
    raw_probabilities = db.Column(db.JSON)
    input_features    = db.Column(db.JSON)
    created_at        = db.Column(db.DateTime,   default=datetime.utcnow)

    VALID_CLASSES: frozenset[str] = frozenset({"optimal", "warning", "critical"})

    def to_dict(self) -> dict:
        """Serialise to a JSON-safe dict."""
        return {
            "prediction_id":     self.prediction_id,
            "device_id":         self.device_id,
            "reading_id":        self.reading_id,
            "model_name":        self.model_name,
            "model_version":     self.model_version,
            "predicted_class":   self.predicted_class,
            "confidence":        float(self.confidence) if self.confidence is not None else None,
            "raw_probabilities": self.raw_probabilities,
            "input_features":    self.input_features,
            "created_at":        self.created_at.isoformat() if self.created_at else None,
        }

    def __repr__(self) -> str:
        return f"<Prediction {self.prediction_id} {self.predicted_class} ({self.confidence})>"

File: greenhouse_aiot/spring-backend/src/main/java/com/aiot/greenhouse/model/Prediction.java
package com.aiot.greenhouse.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.Map;

/**
 * Resultado de inferencia del modelo de IA, almacenado para trazabilidad y análisis.
 */
@Entity
@Table(name = "predictions")
@Getter @Setter @NoArgsConstructor @AllArgsConstructor @Builder
public class Prediction {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "prediction_id")
    @JsonProperty("prediction_id")
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "device_id", nullable = false)
    @JsonIgnoreProperties({"hibernateLazyInitializer", "handler", "registeredBy", "zone"})
    private Device device;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "reading_id", nullable = false)
    @JsonIgnoreProperties({"hibernateLazyInitializer", "handler", "device"})
    private SensorReading reading;

    @Column(name = "model_name", nullable = false, length = 100)
    private String modelName;

    @Column(name = "model_version", nullable = false, length = 20)
    @Builder.Default
    private String modelVersion = "1.0";

    @Enumerated(EnumType.STRING)
    @Column(name = "predicted_class", nullable = false, length = 10)
    @JsonProperty("predicted_class")
    private PredictedClass predictedClass;

    @Column(name = "confidence", precision = 5, scale = 4)
    private BigDecimal confidence;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "raw_probabilities", columnDefinition = "jsonb")
    private Map<String, Double> rawProbabilities;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "input_features", columnDefinition = "jsonb")
    private Map<String, Object> inputFeatures;

    @Column(name = "created_at", nullable = false, updatable = false)
    @Builder.Default
    private LocalDateTime createdAt = LocalDateTime.now();

    /** Clasificación de condición del cultivo. */
    public enum PredictedClass {
        OPTIMAL, WARNING, CRITICAL
    }
}
